File: backend/src/auth/services/auth.service.ts
import {
  BadRequestException,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import * as bcrypt from 'bcrypt';
import { UsersService } from '../../users/services/users.service';
import { LoginDto } from '../dto/login.dto';
import { RegisterDto } from '../dto/register.dto';

const jwtExpiresInSeconds = Number(process.env.JWT_EXPIRES_IN_SECONDS ?? 86400);

@Injectable()
export class AuthService {
  private readonly revokedTokens = new Set<string>();

  constructor(
    private readonly usersService: UsersService,
    private readonly jwtService: JwtService,
  ) {}

  async register(dto: RegisterDto) {
    const existingEmail = await this.usersService.findByEmail(dto.email);
    if (existingEmail) {
      throw new BadRequestException('Email already in use');
    }

    const existingUsername = await this.usersService.findByUsername(dto.username);
    if (existingUsername) {
      throw new BadRequestException('Username already in use');
    }

    const passwordHash = await bcrypt.hash(dto.password, 10);
    const user = await this.usersService.create({
      username: dto.username,
      email: dto.email,
      passwordHash,
      role: 'admin',
    });

    return {
      id: user.id,
      username: user.username,
      email: user.email,
      role: user.role,
      createdAt: user.createdAt,
    };
  }

  async login(dto: LoginDto) {
    const user = await this.usersService.findByEmail(dto.email);
    if (!user) {
      throw new UnauthorizedException('Invalid credentials');
    }

    const passwordMatches = await bcrypt.compare(dto.password, user.passwordHash);
    if (!passwordMatches) {
      throw new UnauthorizedException('Invalid credentials');
    }

    const payload = {
      sub: user.id,
      username: user.username,
      role: user.role,
    };

    const accessToken = await this.jwtService.signAsync(payload, {
      secret: process.env.JWT_SECRET ?? 'change-me-in-production',
      expiresIn: jwtExpiresInSeconds,
    });

    return {
      accessToken,
      tokenType: 'Bearer',
      user: {
        id: user.id,
        username: user.username,
        email: user.email,
        role: user.role,
      },
    };
  }

  logout(token: string) {
    if (!token) {
      throw new BadRequestException('Token is required');
    }

    this.revokedTokens.add(token);
    return { message: 'Logged out successfully' };
  }

  isTokenRevoked(token: string) {
    return this.revokedTokens.has(token);
  }
}

File: backend-java/src/main/java/com/iraninfo/services/AuthService.java
package com.iraninfo.services;

import com.iraninfo.models.User;
import com.iraninfo.utils.JwtUtil;
import org.mindrot.jbcrypt.BCrypt;

import java.sql.SQLException;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public class AuthService {

    private final UserService userService = new UserService();
    private final Set<String> revokedTokens = ConcurrentHashMap.newKeySet();

    public Map<String, Object> register(String username, String email, String password) throws SQLException {
        User existingEmail = userService.findByEmail(email);
        if (existingEmail != null) {
            throw new IllegalArgumentException("Email already in use");
        }

        User existingUsername = userService.findByUsername(username);
        if (existingUsername != null) {
            throw new IllegalArgumentException("Username already in use");
        }

        String passwordHash = BCrypt.hashpw(password, BCrypt.gensalt(10));
        User user = userService.create(username, email, passwordHash, "admin");

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("id", user.getId());
        result.put("username", user.getUsername());
        result.put("email", user.getEmail());
        result.put("role", user.getRole());
        result.put("createdAt", user.getCreatedAt());
        return result;
    }

    public Map<String, Object> login(String email, String password) throws SQLException {
        User user = userService.findByEmail(email);
        if (user == null) {
            throw new SecurityException("Invalid credentials");
        }

        if (!BCrypt.checkpw(password, user.getPasswordHash())) {
            throw new SecurityException("Invalid credentials");
        }

        String accessToken = JwtUtil.generateToken(user.getId(), user.getUsername(), user.getRole());

        Map<String, Object> userMap = new LinkedHashMap<>();
        userMap.put("id", user.getId());
        userMap.put("username", user.getUsername());
        userMap.put("email", user.getEmail());
        userMap.put("role", user.getRole());

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("accessToken", accessToken);
        result.put("tokenType", "Bearer");
        result.put("user", userMap);
        return result;
    }

    public Map<String, Object> logout(String token) {
        if (token == null || token.isEmpty()) {
            throw new IllegalArgumentException("Token is required");
        }
        revokedTokens.add(token);
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("message", "Logged out successfully");
        return result;
    }

    public boolean isTokenRevoked(String token) {
        return revokedTokens.contains(token);
    }
}
